### PR TITLE
feat: namespace-aware tx iterator

### LINKED ISSUE #931: Support tx iterator with namespaces
Blocker for #892 because we have no easy way to iterate through the txs in a block. We need this to test transaction proofs via `verify()`.

From https://github.com/EspressoSystems/espresso-sequencer/issues/892#issuecomment-1878978395 :

> Ugh the impl for `QueryablePayload` needs a complete rewrite for this issue:
> - `len()` is more complex now because we need to sum the number of txs over all namespaces
> - `iter()` is more complex: need to iterate over `([namespace id], [tx index])`

### PR DESCRIPTION
**Pull request by akonring**
closes #931

### CODE REVIEW

**Review by ggutoski — COMMENTED** [review]
Thanks for this! A few notes below.

File: sequencer/src/block2.rs [hunk 1]
```diff
@@ -555,6 +638,104 @@ mod tx_table_entry {
     }
 }
 
+type NsTable = <BlockPayload as hotshot::traits::BlockPayload>::Metadata;
+type TxIndexUnsigned = u32; // need a better name, want `TxIndex` but that's already taken.
+
+struct TxIndex2 {
+    ns_id: VmId,
```
[the remark this answers is not in the document]
Comment (ggutoski) [reply sequencer/src/block2.rs:645]:
```suggestion
    ns_idx: usize,
```

File: sequencer/src/block2.rs [hunk 2]
```diff
@@ -170,6 +171,91 @@ fn get_table_len(table_bytes: &[u8], offset: usize) -> TxTableEntry {
...
+// TODO comment
+fn get_tx_table_len(ns_bytes: &[u8]) -> usize {
+    std::cmp::min(
+        get_table_len(ns_bytes, 0).try_into().unwrap_or(0),
+        (ns_bytes.len() - TxTableEntry::byte_len()) / TxTableEntry::byte_len(),
+    )
+}
+
+// returns (ns_id, payload_offset)
+// payload_offset is not checked, could be anything
```
[the remark this answers is not in the document]
Comment (ggutoski) [reply sequencer/src/block2.rs:191]: It means that this function parses whatever's in the namespace table as an offset. So if that data is corrupted then the returned offset is corrupted.

This should never happen now that the namespace table is part of the header, so consensus nodes enforce a well-formed namespace table. But I want this code / docs to be paranoid.

File: sequencer/src/block2.rs [hunk 3]
```diff
@@ -555,6 +627,83 @@ mod tx_table_entry {
     }
 }
 
+type NsTable = <BlockPayload as hotshot::traits::BlockPayload>::Metadata;
+
+/// TODO do we really need `PartialOrd`, `Ord` here?
+/// Could the `Ord` bound be removed from `QueryablePayload::TransactionIndex`?`
+#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Serialize, Deserialize)]
```
Comment (ggutoski) [inline sequencer/src/block2.rs:634]: @jbearer question for you

File: sequencer/src/block2.rs [hunk 4]
```diff
@@ -555,6 +627,83 @@ mod tx_table_entry {
...
+pub struct TxIterator<'a> {
+    ns_idx: usize, // TODO should be able to eliminate this field
+    ns_iter: Range<usize>,
+    tx_iter: Range<usize>,
+    block_payload: &'a BlockPayload,
+
+    // TODO should be `&'a NsTable`, but that requires upstream change to `QuearyablePayload::iter`:
+    // fn iter<'a>(&'a self, meta: &'a Self::Metadata)
+    //        ++++  ++              ++
+    ns_table: NsTable,
```
Comment (ggutoski) [inline sequencer/src/block2.rs:649]: @jbearer looks like we need another tweak to `QueryablePayload` 😕

**Review by ggutoski — COMMENTED** [review]
I rebased main and force-pushed new commits to this branch. Need approval from @jbearer and/or @nomaxg .

We are currently pushing to close #1009 , #1011 . Some code from this PR will be used to close these subsequent issues.

Code still needs significant tidying. We already have a few issues posted about this #856 #951 #1012 that will be addressed after we close #1011 . In the meantime, I ask that you be forgiving about code cleanliness.

Flagged a couple notes for @jbearer in comments below.

File: sequencer/src/block2.rs (hunk 4 above)
Comment (jbearer) [reply sequencer/src/block2.rs:649]: Ah, that looks like a good change. I can do it, should be quick 🤞

**Review by jbearer — APPROVED** [review]
LGTM

**Pull request merged**